src/test/java/org/broadinstitute/hellbender/tools/spark/transforms/markduplicates/MarkDuplicatesSparkUtilsUnitTest.java

```diff
@@ -0,0 +1,45 @@
+package org.broadinstitute.hellbender.tools.spark.transforms.markduplicates;
+
+import com.google.api.client.util.Lists;
+import com.google.common.collect.ImmutableList;
+import org.broadinstitute.hellbender.utils.test.BaseTest;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+import scala.Tuple2;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+
+public class MarkDuplicatesSparkUtilsUnitTest extends BaseTest {
+    @Test
+    public void testSpanningIterator() {
+        check(Collections.emptyIterator(), Collections.emptyList());
+        check(ImmutableList.of(pair(1, "a")).iterator(),
+                ImmutableList.of(pairIterable(1, "a")));
+        check(ImmutableList.of(pair(1, "a"), pair(1, "b")).iterator(),
+                ImmutableList.of(pairIterable(1, "a", "b")));
+        check(ImmutableList.of(pair(1, "a"), pair(2, "b")).iterator(),
+                ImmutableList.of(pairIterable(1, "a"), pairIterable(2, "b")));
+        check(ImmutableList.of(pair(1, "a"), pair(1, "b"), pair(2, "c")).iterator(),
+                ImmutableList.of(pairIterable(1, "a", "b"), pairIterable(2, "c")));
+        check(ImmutableList.of(pair(1, "a"), pair(2, "b"), pair(2, "c")).iterator(),
+                ImmutableList.of(pairIterable(1, "a"), pairIterable(2, "b", "c")));
+    }
```

Comment:
Could you add some tests that use keys returned by `ReadsKey.keyForRead()`, and include that test case I mentioned above of a sequence of reads with the same name but different read groups? ("A|N", "B|N", "A|N", "B|N")